EKF/control.cpp

```diff
@@ -674,7 +673,31 @@ void Ekf::controlGpsFusion()
 			// We haven't had an absolute position fix for a longer time so need to do something
 			do_reset = do_reset || ((_time_last_imu - _time_last_pos_fuse) > (2 * _params.reset_timeout_max));
 
-			if (do_reset) {
+			// A reset to the EKF-GSF estimate can be performed after a recent takeoff which will enable
+			// recovery from a bad magnetometer or field estimate.
+			// This special case reset can also be requested externally.
+			// The minimum time interval between resets to the EKF-GSF estimate must be limited to
+			// allow the EKF-GSF time to improve its estimate if the first reset was not successful.
+			const bool stopped_following_gps_velocity = (_time_last_imu - _time_last_vel_fuse) > _params.EKFGSF_reset_delay &&
+							   (_time_last_vel_fuse > _time_last_on_ground_us);
+			const bool recent_takeoff = (_control_status.flags.in_air && (_imu_sample_delayed.time_us - _time_last_on_ground_us) < 30000000);
+			const bool reset_yaw_to_EKFGSF = (do_reset || _do_emergency_yaw_reset || stopped_following_gps_velocity) && recent_takeoff && ((_imu_sample_delayed.time_us - _emergency_yaw_reset_time) > 5000000);
```

Comment:
A loss of navigation after takeoff is an emergency. However the 30 second activation limit can be removed when we get more confidence in the long term stability of the algorithm. Preliminary test result are encouraging, but we haven't done extended FW flights or hovering flights in one position yet.

Reply:
But could there be a situation where we need to do the reset in flight? In this case the variable `recent_takeoff` will not allow for the EKFGSF reset.

Reply:
I think keeping the concept of a defined 'post takeoff' time period makes sense in the context of setting loss of navigation parameters when we are trying to detect loss of navigation caused by a bad yaw estimate. This is because crash logs submitted in the past have shown a loss of navigation immediately after takeoff due to a bad yaw estimate is a common cause and is exacerbated by proximity to ground obstacles during that period of operation. Recovering rapidly from this type of error was the primary motivator for developing this algorithm.

Usually if the EKF takes off with a good yaw angle, it is able to reject the magnetometer or learn the new biases and field values for errors that develop later in flight, but not always. There are yaw induced failure scenarios that can occur later in flight that we may also be able to detect abd recover from using this estimator.

Splitting the detection into a period immediately following takeoff with a more rapid time response and a subsequent period with a slower time response and lower false detection rate will be investigated.

Reply:
Ok, I understand now. The different response time is a good point